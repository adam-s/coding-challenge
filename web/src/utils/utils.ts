export const removeArrayDuplicates = <T, U extends keyof T>(
  B: T[],
  A: T[],
  key: U
): T[] => {
  const ids = new Set(A.map((d) => d[key]));
  return [...B.filter((d) => !ids.has(d[key])), ...A];
};

const isNumeric = (value: unknown): value is number => {
  return typeof value === 'number';
};

const isString = (value: unknown): value is string => {
  return typeof value === 'string';
};

export const sortByKey = <T, U extends keyof T>(
  A: T[],
  key: U,
  order: 'ASC' | 'DESC' = 'ASC'
): T[] => {
  const sorting = order === 'ASC' ? 1 : -1;
  return A.sort((a1: any, a2: any) => {
    if (!a1) return 1;
    if (!a1[key]) return 1;
    if (!a2) return -1;
    if (!a2[key]) return 1;
    if (isNumeric(a1[key]) && isNumeric(a2[key])) {
      return a1[key] - a2[key] * sorting;
    }
    if (isString(a1[key]) && isString(a2[key])) {
      return (
        a1[key].toLowerCase().localeCompare(a2[key].toLowerCase()) * sorting
      );
    }
    return 0;
  });
};

// https://codesandbox.io/s/draggable-material-ui-oj3wz?file=/src/helpers.ts:28-325
// a little function to help us with reordering the result [Don't need this for our use case]
export const reorder = <T>(
  list: T[],
  startIndex: number,
  endIndex: number
): T[] => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

export const insureSorted = <T>(A: T[], key: string) =>
  A.map<T>((a, index) => Object.assign({}, a, { [key]: index }));
