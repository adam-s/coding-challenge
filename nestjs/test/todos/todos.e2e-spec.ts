import * as request from 'supertest';
import { APP_URL } from '../utils/constants';

const GQL_ENDPOINT = '/graphql';

describe('Todos (e2e)', () => {
  it('should create and return a todo', () => {
    const todo = {
      id: '7',
      name: 'first todo',
      completed: false,
      sort: 0,
    };
    return request(APP_URL)
      .post(GQL_ENDPOINT)
      .send({
        query: `mutation {createTodo(newTodoData: {name: "${todo.name}", sort: ${todo.sort}, completed: ${todo.completed}}) {id, name, completed, sort}}`,
      })
      .expect(200)
      .expect((res) => {
        expect(res.body.data.createTodo).toEqual(todo);
      });
  });

  it('should update and return a todo', () => {
    const todo = {
      id: '7',
      name: 'first todo',
      completed: false,
      sort: 0,
    };

    const name = 'updated todo';

    return request(APP_URL)
      .post(GQL_ENDPOINT)
      .send({
        query: `mutation {updateTodo(updateTodoData: {id: "${todo.id}", name: "${name}", sort: ${todo.sort}, completed: ${todo.completed}}) {id, name, completed, sort}}`,
      })
      .expect(200)
      .expect((res) => {
        expect(res.body.data.updateTodo.name).toEqual(name);
      });
  });

  it('should find and return a todo by id', () => {
    const todo = {
      id: '7',
      name: 'updated todo',
      completed: false,
      sort: 0,
    };

    return request(APP_URL)
      .post(GQL_ENDPOINT)
      .send({
        query: `query {
          todo(id: "${todo.id}") { id, name, completed, sort } 
        }`,
      })
      .expect(200)
      .expect((res) => {
        expect(res.body.data.todo).toEqual(todo);
      });
  });

  it('should find and return all todos', () => {
    return request(APP_URL)
      .post(GQL_ENDPOINT)
      .send({
        query: `query {
          todos { data { id, name, completed, sort } }
        }`,
      })
      .expect(200)
      .expect((res) => {
        expect(res.body.data.todos).toMatchSnapshot();
      });
  });
});
