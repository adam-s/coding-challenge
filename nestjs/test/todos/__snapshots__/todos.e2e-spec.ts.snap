// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos (e2e) should find and return all todos 1`] = `
Object {
  "data": Array [
    Object {
      "completed": false,
      "id": "1",
      "name": "Get all contingencies squared away",
      "sort": 0,
    },
    Object {
      "completed": false,
      "id": "2",
      "name": "Clear the title",
      "sort": 1,
    },
    Object {
      "completed": false,
      "id": "3",
      "name": "Get final mortgage approval",
      "sort": 2,
    },
    Object {
      "completed": false,
      "id": "4",
      "name": "Review your closing disclosure",
      "sort": 3,
    },
    Object {
      "completed": false,
      "id": "5",
      "name": "Do a final walk-through",
      "sort": 4,
    },
    Object {
      "completed": false,
      "id": "6",
      "name": "Bring the necessary documentation to closing",
      "sort": 5,
    },
    Object {
      "completed": false,
      "id": "7",
      "name": "updated todo",
      "sort": 0,
    },
  ],
}
`;
